pipeline/util/urllib.py

```diff
@@ -36,39 +39,47 @@ def download(
     Note: timestamps of remote files are not retained.
     """
 
-    with tempfile.TemporaryDirectory() as tmpdirname:
-        tmpfilename = os.path.join(tmpdirname, filename)
-        finalfilename = os.path.join(destination_path, filename)
-
-        if not os.path.exists(destination_path):
-            os.makedirs(destination_path)
-
-        headers = {'User-Agent': user_agent}
-        if not force and os.path.exists(finalfilename):
-            pre_existing_file_last_modified = datetime.utcfromtimestamp(os.path.getmtime(finalfilename))
-            headers['If-Modified-Since'] = pre_existing_file_last_modified.strftime('%a, %d %b %Y %H:%M:%S GMT')
-
-        response = get(source, timeout=timeout, headers=headers, stream=True)
-        if response.status_code == 304:
-            # File not modified since last download
-            return
-
-        with open(tmpfilename, 'wb') as file:
-            for chunk in response.iter_content(chunk_size=1024):
-                if chunk:
-                    file.write(chunk)
-
-        last_modified = None
-        last_modified_str = response.headers.get('Last-Modified')
-        if last_modified_str:
-            last_modified = parsedate_to_datetime(last_modified_str)
-            os.utime(tmpfilename, (last_modified.timestamp(), last_modified.timestamp()))
-
-        if gzip:
-            with open(tmpfilename, 'rb') as f_in, gzip.open(tmpfilename + '.gz', 'wb') as f_out:
-                shutil.copyfileobj(f_in, f_out)
-                if last_modified:
-                    os.utime(tmpfilename + '.gz', (last_modified.timestamp(), last_modified.timestamp()))
-                os.replace(tmpfilename + '.gz', finalfilename + '.gz')
-
-        os.replace(tmpfilename, finalfilename)
+    final_filename = Path(destination_path) / filename
+    tmp_filename = final_filename.parent / (final_filename.name + '.tmp')
+
+    if not final_filename.parent.exists():
+        final_filename.parent.mkdir(parents=True)
+
+    headers = {'User-Agent': user_agent}
+    if not force and final_filename.exists():
+        pre_existing_file_last_modified = datetime.utcfromtimestamp(final_filename.stat().st_mtime)
+        headers['If-Modified-Since'] = pre_existing_file_last_modified.strftime('%a, %d %b %Y %H:%M:%S GMT')
```

Comment:
Note that [GMT is not UTC](https://24timezones.com/gmt-vs-utc):

> Moreover, to maintain the consistent time system, UTC does not observe Daylight Saving Time (unlike GMT).

So, if I understand the code right, a) `If-Modified-Since` is shifted during DST, and b) conditional requests break unnoticably during the standard time <-> DST shift.

Reply:
Good spot, thanks. Switched to UTC with 20f8cce0368225750a0be70aed53758cdbc48e68